tidy3d/plugins/adjoint/components/geometry.py

```diff
@@ -0,0 +1,435 @@
+                return Et, En, Ez
+
+            # get forward and adjoint fields in edge basis
+            e_t_fwd, e_n_fwd, e_z_fwd = edge_basis(grad_data_fwd, components_xyz=("Ex", "Ey", "Ez"))
+            e_t_adj, e_n_adj, e_z_adj = edge_basis(grad_data_adj, components_xyz=("Ex", "Ey", "Ez"))
+
+            # get displacement fields
+            _, eps_n, _ = edge_basis(grad_data_eps, components_xyz=("eps_xx", "eps_yy", "eps_zz"))
+            d_n_fwd = eps_n * e_n_fwd
+            d_n_adj = eps_n * e_n_adj
+
+            # TODO: actually compute these..
+            delta_eps_12 = 1.0
+            delta_eps_inv_12 = 1.0
+
+            def scalar_integrand(s: float, z: float) -> float:
+
+                x, y = r(s)
+                coords_interp = dict(x=x, y=y, z=z)
+
+                def evaluate(scalar_field) -> float:
+                    scalar_field = scalar_field.isel(f=0)
+                    scalar_field = scalar_field.interp(**coords_interp)
+                    return scalar_field
+
+                contrib_e_t = evaluate(delta_eps_12 * e_t_fwd * e_t_adj)
+                contrib_e_z = evaluate(delta_eps_12 * e_z_fwd * e_z_adj)
+                contrib_d_n = -evaluate(delta_eps_inv_12 * d_n_fwd * d_n_adj)
+                return s * (contrib_e_t + contrib_d_n + contrib_e_z)
+
+            slab_min, slab_max = self.slab_bounds
+            length_axis = slab_max - slab_min
+
+            num_cells_edge = int(length_edge * PTS_PER_WVL_INTEGRATION / wvl_mat) + 1
+            num_cells_axis = int(length_axis * PTS_PER_WVL_INTEGRATION / wvl_mat) + 1
+
+            s_vals = np.linspace(0, 1, num_cells_edge)
+            z_vals = np.linspace(slab_min, slab_max, num_cells_axis)
+            ds = np.mean(np.diff(s_vals)) * length_edge
+            dz = np.mean(np.diff(z_vals))
+
+            integrand = 0.0
+            for s in s_vals:
+                integrand += np.sum(scalar_integrand(s=s, z=z_vals).fillna(0).values)
+
+            return n * np.real(ds * dz * integrand)
+
+        def mod(num: int) -> int:
+            """Get index modulo number of vertices."""
+            return num % len(self.vertices)
+
+        edge_contributions = []
+        for i_vertex, vertex in enumerate(self.vertices):
+            i_next = mod(i_vertex + 1)
+            vertex_next = self.vertices[i_next]
+            edge_contribution = edge_contrib(vertex, vertex_next, axis=self.axis)
+            edge_contributions.append(edge_contribution)
+
+        new_vertices = []
+        for edge_index, edge_i in enumerate(edge_contributions):
+            prev_edge_index = mod(edge_index - 1)
+            edge_i_m1 = edge_contributions[prev_edge_index]
+            vertex_contrib = tuple(val_i - val_i_m1 for val_i, val_i_m1 in zip(edge_i, edge_i_m1))
+            new_vertices.append(vertex_contrib)
+
+        return self.copy(update=dict(vertices=new_vertices))
+
+
+JaxGeometryType = Union[JaxBox, JaxPolySlab]
```

Comment:
Sorry if I missed this, but did you mention that in pre/1.9.0 you'll only support boxes? If so, maybe we shouldn't give the user access to poly slabs here?

Reply:
Yea, this is just for internal testing right now. Before I clean everything up for the pre-release, I'll be sure to remove JaxPolySlab from this and other places where it can be imported.